### ISSUE TITLE
CRIU pre-copy + post-copy

### ISSUE THREAD

**Issue by shubham95** [checkpoint-restore/criu#1095]
In Hybrid copy
[https://lisas.de/~adrian/posts/2016-Oct-14-combining-pre-copy-and-post-copy-migration.html](url)

when we do last dump on source and post-copy on destination.
How restore work in this case. it will handle the page fault of dirty pages through lazy pages 
But how it copy non dirty pages. does it copy non dirty pages from pre-dumps file on the fly because post copy start without address space ?

**Comment by adrianreber** [comment]
> But how it copy non dirty pages. does it copy non dirty pages from pre-dumps file on the fly

Yes, CRIU uses the pages from the pre-dump and no, dirty pages are not involved during lazy migration.

> because post copy start without address space ?

I don't understand that part. How does this part of the question relate to the first part.

CRIU will restore the process with all the memory pages which are available on the destination system. In the article above this pages have been transferred using pre-dumps.

As we tell CRIU in that article to mark all missing pages as being handled by usefaultd, the kernel waits for those pages to be inserted into the kernel when it hits a missing page.

I am confused how you use dirty pages in combination with lazy migration. Dirty pages are related to pre-copy. During post-copy/lazy migration using userfaultd we do not have dirty pages any more.

Dirty pages are only relevant during pre-dump. In the first pre-dump all pages are written to disk or page-server, in a following dump, only dirty pages are dumped. Dirty pages are memory pages which have changed since the last dump.

**Comment by shubham95** [comment]
Let assume at each pre-dump no. of dirty pages decreases
1.    Pre-dump : 2G
2.   Pre-dump :  1G
3.   Pre-dump :   200MB
4.   Now we are going to do post-copy/ lazy dump and the dirty pages are of size 100MB

At Destination, we have all 3 pre-dumps file 
and after final dump we call criu restore

Now there are two ways to restore 
1. restorer doesn't start process . **it first copy pages from all pre-dump files** and start the process now there are 100MB of pages which are not here for that it will copy pages by generating page-fault(uffd)

2. restorer starts the process at destination without copy any pages in process address space. when page fault occur  either it will handled from pre-dump pages or request is sent to source


The Diffrence between 1st and 2nd approach is 
In 1st case downtime is more because we are copying the pages from pre-dump and process is stopped till that time (and it is not serving the purpose of post-copy)
while in 2nd case many pagefault is handled locally so it is not as bad as normal post-copy

PS: 
I am trying to understand the parsing of all image files. In which file do I have to look .
In which function we are parsing the pages-1.img file

**Comment by adrianreber** [comment]
Okay, I think I do understand your question now better.

In theory the downtime of approach 2 is indeed shorter and the process will start to run earlier with approach 2. But the actual downtime will be much bigger, because the process has to stop all the time and wait for a page to be available via userfaultfd. So process is stopped, a message goes to user-space via userfaultfd and until user-space answers that request the process is stopped. That is slow and can take a very long time. Copying all pre-dumped pages at right location like in approach 1 is much faster for the overall downtime.

So in approach 2 your process will start faster, but it will take much more time until it can do the task it is supposed to do. The more pages you can restore via pre-copy, the better.

**Comment by shubham95** [comment]
thanks @adrianreber 
So how criu restore works which approach it follows in hybrid migration ?

**Comment by adrianreber** [comment]
Can you rephrase? Not sure I understand your question.

Am 6. Juni 2020 12:09:54 MESZ schrieb SHUBHAM PANDEY <notifications@github.com>:
>thanks @adrianreber 
>So how criu restore works which approach it follows in hybrid migration
>?

**Comment by shubham95** [comment]
In this article
[https://lisas.de/~adrian/posts/2016-Oct-14-combining-pre-copy-and-post-copy-migration.html](url)

it is mentioned that we can use pre-dump+postcopy

So, After pre-dumps during final post-copy 
How restorer works on destination Does it follow approach 1 or 2 (mention above)

**Comment by rppt** [comment]
Hi @shubham95,

> So, After pre-dumps during final post-copy
> How restorer works on destination Does it follow approach 1 or 2 (mention above)

The restorer (+lazy-pages) follow the 1 approach. First, the restorer populates the process with the pages from the pre-copy, then starts it and the lazy-pages daemon fills in the missing memory (the last 100M in your example).

**Comment by shubham95** [comment]
Thanks @rppt and @adrianreber  for clarifying this.
